snuba/query/processors/hash_map_optimizer.py

```diff
@@ -0,0 +1,192 @@
+from enum import Enum
+
+from snuba import environment
+from snuba.clickhouse.processors import QueryProcessor
+from snuba.clickhouse.query import Query
+from snuba.clickhouse.translators.snuba.mappers import (
+    KEY_COL_MAPPING_PARAM,
+    KEY_MAPPING_PARAM,
+    TABLE_MAPPING_PARAM,
+    mapping_pattern,
+)
+from snuba.query.conditions import (
+    BooleanFunctions,
+    get_first_level_and_conditions,
+    get_first_level_or_conditions,
+)
+from snuba.query.expressions import Column, Expression
+from snuba.query.expressions import FunctionCall as FunctionExpr
+from snuba.query.expressions import Literal as LiteralExpr
+from snuba.query.matchers import Any, FunctionCall, Literal, Or, Param, String
+from snuba.request.request_settings import RequestSettings
+from snuba.state import get_config
+from snuba.utils.metrics.backends.wrapper import MetricsWrapper
+
+metrics = MetricsWrapper(environment.metrics, "processors.tags_hash_map")
+
+
+class ConditionClass(Enum):
+    IRRELEVANT = 1
+    OPTIMIZABLE = 2
+    NOT_OPTIMIZABLE = 3
+
+
+class HashMapOptimizer(QueryProcessor):
+    """
+    Optimize tags conditions by relying on the tags_hash_map column.
+    It transforms tags conditions that are in the form of
+    `tags.value[indexOf(tags.key, 'my_tag')] = 'my_val'`
+    into
+    `has(_tags_hash_map, cityHash64('my_tag=my_val'))`
+
+    Supported use case:
+    - direct equality. Example above
+    - tags expression nested into ifNull conditions like:
+      `ifNull('tags.value[indexOf(tags.key, 'my_tag')]', '') = ...`
+    - tags conditions in both where and having
+
+    Unsupported use cases:
+    - everything that cannot be checked through the tags hash map
+      like !=, LIKE, IS NULL
+    - it will not optimize a condition if another condition still
+      requires unpacking the tags column. Like
+      `tags[a] = b AND tags[b] LIKE 'asd'`
+      That would load an additional column for almost no gain thus
+      actually degrading performance.
+    - `ifNull('tags.value[indexOf(tags.key, 'my_tag')]', '') = ''`
+       this condition is equivalent to looking whether a tag is
+       missing, which cannot be done with the hash map.
```

Comment:
Why can't it be done with the hash map? Doesn't the absence of the tag key in the _tags_hash_map prove that it must be null?

Reply:
This is actually more of a hashset I think I should change the name. Where each element is `cityHash64('key=value')` You cannot run a query to check if a given key is there or not.